kubeadm: Add support for using an external CA whose key is never stored in the cluster

nckturner opened the pull request:
We allow a kubeadm user to use an external CA by checking to see if ca.key is missing and skipping cert checks and kubeconfig generation if ca.key is missing.  We also pass an empty arg --cluster-signing-key-file="" to kube controller manager so that the csr signer doesn't start. 

<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
-->

**What this PR does / why we need it**:

This PR allows the kubeadm certs phase and kubeconfig phase to be skipped if the ca.key is missing but all other certs are present.  

**Which issue this PR fixes** : 

Fixes kubernetes/kubeadm/issues/280

**Special notes for your reviewer**:

@luxas @mikedanese @fabriziopandini 

**Release note**:
<!--  Steps to write your release note:
1. Use the release-note-* labels to set the release note state (if you have access)
2. Enter your extended release note in the below block; leaving it blank means using the PR title as the release note. If no release note is required, just write `NONE`.
-->
```release-note
kubeadm: Add support for using an external CA whose key is never stored in the cluster
```

dims commented:
/ok-to-test

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -233,14 +234,19 @@ func (i *Init) Run(out io.Writer) error {
+		}
+
+	} else {
+		fmt.Printf("[externalca] No ca.key detected, assuming external CA.  Skipping certs and kubeconfig.\n")
```
nit: `fmt.Println`
nit: `No ca.key detected, but all other files generated from it are available, so using external CA mode. Won't generate certs nor kubeconfig files` (or something like that)

nckturner replied:
Fixed

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -193,18 +194,25 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion *versio
 
+	// If using external CA, pass empty string to controller manager instead of ca.key path, so that the csrsigning controller fails to start
+	if certphase.UsingExternalCA(cfg) {
+		defaultArguments["cluster-signing-key-file"] = ""
```
for clarity; I'd always set `cluster-signing-key-file` and `cluster-signing-cert-file` in `defaultArguments`
and in this special case; unset both (although one only is technically needed, it's safest to reset both of them

nckturner replied:
Makes sense, fixed.

luxas commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+		return false
+	}
+
+	if !ValidatePrivateKey(cfg.CertificatesDir, kubeadmconstants.ServiceAccountKeyBaseName, "service account") {
```
What about `sa.pub`? can we make a case for that one as well?

nckturner replied:
I added a case, to superficially check for its existence, tell me what you think.

luxas commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+	return true
+}
+
+func ValidateCACert(pkiDir string, baseName string, UXName string) bool {
```
Please add a comment for all these

nckturner replied:
Done

luxas commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+	return true
+}
+
+func ValidateCACertAndKey(pkiDir string, baseName string, UXName string) bool {
```
And unit test all these new funcs you are creating.

nckturner replied:
Done, though tell me if I should refactor them into table driven tests.

luxas replied:
Table driven tests are common practice, please write such ones.
Nearly all unit tests in `cmd/kubeadm` are table driven

luxas reviewed (commented):
Basically LGTM; some asks and nits + unit (which I know you're working on already :+1:)

Thanks for this contribution!

luxas commented:
From the verify job:
```
I0817 11:02:41.808] Run ./hack/update-bazel.sh
...
W0817 11:20:11.506] Errors from golint:
W0817 11:20:11.507] cmd/kubeadm/app/phases/certs/certs.go:162:1: exported function ValidateCACert should have comment or be unexported
W0817 11:20:11.510] cmd/kubeadm/app/phases/certs/certs.go:178:1: exported function ValidateCACertAndKey should have comment or be unexported
W0817 11:20:11.510] cmd/kubeadm/app/phases/certs/certs.go:191:1: exported function ValidateSignedCert should have comment or be unexported
W0817 11:20:11.511] cmd/kubeadm/app/phases/certs/certs.go:214:1: exported function ValidatePrivateKey should have comment or be unexported
```
Please fix.

luxas commented:
This definitely deserves a release note :smile:

mikedanese commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+	return true
+}
+
+func ValidateSignedCert(pkiDir string, CABaseName string, baseName string, UXName string) bool {
```
don't start go variables with capitals. caBaseName...

nckturner replied:
Fixed

fabriziopandini commented:
:+1

mikedanese commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+		return false
+	}
+
+	if !ValidateSignedCert(cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, kubeadmconstants.APIServerCertAndKeyBaseName, "API server") {
```
make this table driven. https://play.golang.org/p/UJY4McemtD

nckturner replied:
Can you explain what you are suggesting a little more?  Do you mean this entire function?

mikedanese commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -120,6 +122,105 @@ func NewFrontProxyClientCertAndKey(frontProxyCACert *x509.Certificate, frontProx
+	// Check CA Cert
+	caCert, err := pkiutil.TryLoadCertFromDisk(pkiDir, baseName)
+	if err != nil {
+		fmt.Printf("failure loading certificate for %s: %v", UXName, err)
```
raw prints without linebreaks are going to look pretty funky in the output.

nckturner replied:
Fixed

fabriziopandini commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -233,14 +234,19 @@ func (i *Init) Run(out io.Writer) error {
-	if err := cmdphases.CreatePKIAssets(i.cfg); err != nil {
-		return err
-	}
+	if !certphase.UsingExternalCA(i.cfg) {
```
According to the issue, it is expected to have also required kubeconfig files on disk. Is kubeadm going to test this precondition?

luxas replied:
Yes, that should also be checked, good catch

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -277,14 +277,21 @@ func (i *Init) Run(out io.Writer) error {
-		return err
-	}
+	if res, err := certsphase.UsingExternalCA(i.cfg); !res {
+		fmt.Printf("[externalca] Not using external CA mode: %v\n", err)
```
I don't think we should print anything here, this is the "normal" state anyway...

nckturner replied:
Removed the print, if someone wants to see why external CA mode is not active they would have to add print statements to this step.  (a logger with verbosity levels would be nice but it seems to be not used in kubeadm)

luxas commented on cmd/kubeadm/app/cmd/init.go:
```diff
@@ -277,14 +277,21 @@ func (i *Init) Run(out io.Writer) error {
-	if err := certsphase.CreatePKIAssets(i.cfg); err != nil {
-		return err
-	}
+	if res, err := certsphase.UsingExternalCA(i.cfg); !res {
```
err seems to be ignored here?

nckturner replied:
I'd say its not an error at this step, so I changed it to be a string representing the reason that we are not in "Using External CA".

luxas commented on cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go:
```diff
@@ -200,6 +200,38 @@ func TryLoadKeyFromDisk(pkiPath, name string) (*rsa.PrivateKey, error) {
+	}
+
+	var p *rsa.PublicKey
+	p = pubKeys[0].(*rsa.PublicKey)
```
Any reason to not use := here?

nckturner replied:
Fixed.

luxas commented on cmd/kubeadm/app/phases/certs/pkiutil/pki_helpers.go:
```diff
@@ -200,6 +200,38 @@ func TryLoadKeyFromDisk(pkiPath, name string) (*rsa.PrivateKey, error) {
+
+	var key *rsa.PrivateKey
+	// Allow RSA format only
+	if k, ok := privKey.(*rsa.PrivateKey); ok {
```
nit on style, this can be moved out of the if clause and you can check for `!ok`, return in that case, otherwise just use `:=` afterwards.

Generally I'm trying to minimize the amount of indentation in funcs and avoid if/else when possible. This is still a nit though ;)

nckturner replied:
Fixed

luxas reviewed (commented):
Left some nits, please fix at least most of them. Other than that, LGTM

luxas commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -398,6 +400,110 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
+func UsingExternalCA(cfg *kubeadmapi.MasterConfiguration) (bool, string) {
+
+	if err := validateCACert(certKeyLocation{cfg.CertificatesDir, kubeadmconstants.CACertAndKeyBaseName, "", "CA"}); err != nil {
+		return false, err.Error()
```
I actually preferred returning an error like you did in the earlier version, but I'm not blocking this PR on that

luxas commented on cmd/kubeadm/app/phases/certs/certs.go:
```diff
@@ -398,6 +400,110 @@ func writeKeyFilesIfNotExist(pkiDir string, baseName string, key *rsa.PrivateKey
+		return err
+	}
+
+	_, err := pkiutil.TryLoadKeyFromDisk(l.pkiDir, l.caBaseName)
```
Not needed for this PR; but in some future PR we should make sure the public key in the cert and key match each other (we have other places that do similar functionality like this so...)

luxas commented on cmd/kubeadm/app/phases/controlplane/manifests.go:
```diff
@@ -215,6 +217,13 @@ func getControllerManagerCommand(cfg *kubeadmapi.MasterConfiguration, k8sVersion
+	// If using external CA, pass empty string to controller manager instead of ca.key/ca.crt path,
+	// so that the csrsigning controller fails to start
+	if res, _ := certphase.UsingExternalCA(cfg); res {
+		defaultArguments["cluster-signing-key-file"] = ""
```
Gah, should have spotted this earlier, I realize `delete(defaultArguments, "cluster...")` is probably preferable over this method... WDYT?

It would be great to get an unit test for this as well if possible...

nckturner replied:
The nice thing about setting to "" is that (in controller manager logs) you get:

> Skipping "csrsigning"

instead of: 

> Failed to start certificate controller: error reading CA cert file "/etc/kubernetes/ca/ca.pem": open /etc/kubernetes/ca/ca.pem: no such file or directory

when using `delete()` which I think is slightly better, but I will switch them if you prefer.

luxas replied:
With that context (thanks for reminding me about the UX), it makes sense to pass `""`

luxas reviewed (commented):
WDYT about the `delete(...` comment?

luxas reviewed (approved):
Thanks for this!

/lgtm

Commit:
Adds check for external CA

We allow a kubeadm user to use an external CA by checking to see if ca.key is missing and skipping cert checks and kubeconfig generation if ca.key is missing.

nckturner commented:
@luxas @mikedanese I squashed my commits, can I get another lgtm?

mikedanese commented:
/lgtm

mikedanese commented:
/retest

mikedanese commented:
/retest

fabriziopandini commented:
🎉 well done!
